desktop/electronDownloadManager.ts

```diff
@@ -0,0 +1,165 @@
+/**
+ * This file is a replicated version of the `electron-dl` package.
+ * It provides a download manager for Electron applications.
+ * The package simplifies the process of downloading files in Electron apps
+ * by providing a high-level API and handling various download-related tasks.
+ * We decided to replicate the functionality of the `electron-dl` package for easier maintenance.
+ * More context: https://github.com/Expensify/App/issues/35189#issuecomment-1959681109
+ */
+import type {BrowserView, BrowserWindow, DownloadItem, Event, Session} from 'electron';
+import {app, shell} from 'electron';
+import * as path from 'path';
+import type {Options} from './electronDownloadManagerType';
+
+/**
+Error thrown if `item.cancel()` was called.
+*/
+class CancelError extends Error {}
+
+/**
+ * Returns the filename with extension based on the given name and MIME type.
+ * @param name - The name of the file.
+ * @param mime - The MIME type of the file.
+ * @returns The filename with extension.
+ */
+const getFilenameFromMime = (name: string, mime: string): string => {
+    const extensions = mime.split('/').pop();
+    return `${name}.${extensions}`;
+};
+
+/**
+ * Registers a listener for download events in Electron session.
+ *
+ * @param session - The Electron session to register the listener on.
+ * @param options - The previous options for the download manager.
+ * @param callback - The callback function to be called when a download event occurs.
+ */
+const registerListener = (session: Session, options: Options, callback: (error: Error | null, item?: DownloadItem) => void = () => {}): void => {
+    const downloadItems = new Set<DownloadItem>();
```

Comment:
The register listener would only be called once the queue is processing an item. As such, I don't see a possibility that we'll have multiple `downloadItems` by any chance. It might happen that the item that we have received in this callback is different from what the `registerListener` was registered for. For this case, we need to check if the item to be `downloaded` is the same as the item that queue is processing right now.